lib/src/dsl_util.rs

```diff
@@ -503,3 +616,92 @@ where
         .dedup()
         .collect()
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_expect_arguments() {
+        fn empty_span() -> pest::Span<'static> {
+            pest::Span::new("", 0, 0).unwrap()
+        }
+
+        fn function(
+            name: &'static str,
+            args: impl Into<Vec<ExpressionNode<'static, u32>>>,
+            keyword_args: impl Into<Vec<KeywordArgument<'static, u32>>>,
+        ) -> FunctionCallNode<'static, u32> {
+            FunctionCallNode {
+                name,
+                name_span: empty_span(),
+                args: args.into(),
+                keyword_args: keyword_args.into(),
+                args_span: empty_span(),
+            }
+        }
+
+        fn value(v: u32) -> ExpressionNode<'static, u32> {
+            ExpressionNode::new(v, empty_span())
+        }
+
+        fn keyword(name: &'static str, v: u32) -> KeywordArgument<'static, u32> {
+            KeywordArgument {
+                name,
+                name_span: empty_span(),
+                value: value(v),
+            }
+        }
+
+        let f = function("foo", [], []);
+        assert!(f.expect_no_arguments().is_ok());
+        assert!(f.expect_some_arguments::<0>().is_ok());
+        assert!(f.expect_arguments::<0, 0>().is_ok());
+        assert!(f.expect_named_arguments::<0, 0>(&[]).is_ok());
+
+        let f = function("foo", [value(0)], []);
+        assert!(f.expect_no_arguments().is_err());
+        assert_eq!(
+            f.expect_some_arguments::<0>().unwrap(),
+            (&[], [value(0)].as_slice())
+        );
+        assert_eq!(
+            f.expect_some_arguments::<1>().unwrap(),
+            (&[value(0)], [].as_slice())
+        );
+        assert!(f.expect_arguments::<0, 0>().is_err());
+        assert_eq!(f.expect_arguments::<1, 1>().unwrap(), (&[value(0)], [None]));
+        assert!(f.expect_named_arguments::<0, 0>(&[]).is_err());
+        assert_eq!(
+            f.expect_named_arguments::<1, 0>(&["a"]).unwrap(),
+            ([&value(0)], [])
+        );
+
+        let f = function("foo", [], [keyword("a", 0)]);
+        assert!(f.expect_no_arguments().is_err());
+        assert!(f.expect_some_arguments::<1>().is_err());
+        assert!(f.expect_arguments::<1, 0>().is_err());
+        assert!(f.expect_named_arguments::<0, 0>(&[]).is_err());
+        assert!(f.expect_named_arguments::<1, 0>(&[]).is_err());
+        assert_eq!(
+            f.expect_named_arguments::<1, 0>(&["a"]).unwrap(),
+            ([&value(0)], [])
+        );
+        assert_eq!(
+            f.expect_named_arguments::<1, 1>(&["a", "b"]).unwrap(),
+            ([&value(0)], [None])
+        );
+        assert!(f.expect_named_arguments::<1, 1>(&["b", "a"]).is_err());
+
+        let f = function("foo", [value(0)], [keyword("a", 1), keyword("b", 2)]);
+        assert!(f.expect_named_arguments::<0, 0>(&[]).is_err());
+        assert!(f.expect_named_arguments::<1, 1>(&["a", "b"]).is_err());
```

Comment:
Can we add a test for ["b", "a"] (e.g. input order does not match programmer order)?  Is this expected to permute the input spans to match the specified keyword order?

Reply:
I think you mean `[_, "b", "a"]`. Added tests. The order of the returned array should follow the `names` arguments.